packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridDataSourceLazyLoader.ts

```diff
@@ -299,6 +309,12 @@ export const useGridDataSourceLazyLoader = (
             delete dataRowIdToModelLookup[rowId];
             duplicateRowCount += 1;
           }
+
+          // re-open the detail panel if it was closed by this hook
+          if (closedDetailPanels.current.has(rowId)) {
+            expandedDetailPanels.add(rowId);
```

Comment:
Is it enough? Shouldn't it get updated via `setState`?
Directly mutating the state would cause store binding subscribed to the state slice to have the outdated value, resulting in unnecessary side effects.

How about cloning the original `expandedDetailPanels` and replacing the previous one in the state using `setState` after updates?
(Same would be true for the other mutation in `addSkeletonRows()`)